readme.txt
The Binaural Engine allows to put several audio sources, reflecting walls, and microhpones in
threedimensional space and calculates the spreading of the sound waves. When configured right,
it allows you to perceive real depth of sound when you listen to the result via headphones.

The executable needs a text file as parameter, which can contain the data listed and described below.

At the current state it can only process and output a raw audio format (for input files: mono, 32 bit
floating point values, output files can have two channels [see below]).

I recommend using audacity for conversion, Download at: https://www.audacityteam.org/
To import a resulting RAW file, go to File -> Import -> Raw Data. Select your file and open it. Select
'32-bit float' and '2 Channels (Stereo)' or '1 Channel (Mono)', depending on the content of the
sound file.
If you want to convert an MP3, WAV (etc) to a RAW file the engine can process, open it in audacity,
convert it to mono (Click on the triangle on the left of your track and then 'Split Stereo to Mono'),
delete one of the mono tracks, select the part you want do use in the Binaural Engine and then
File->Export Selected Audio. Select 'Other uncompressed files', below on 'RAW (header-less)'
and 32-bit-float, then save.

Use the GUI editor to configure your scene and select your raw input and output files. The editor
saves a file (*.b3s = binaural 3D scene) and launches the filter that does the calculations.
The content of the file is described below:

------ The following data can be put into the file that is processed by the filter ------

The "speaker" plays a sound file at a threedimensional position in a desired volume. The position is
measured in meters (like all other position values). The start delay is optional as well as the
keyframes (see below).
Flags are: 1 for "speaker disabled", 2 for "no direct pass", 4 for "no reflection pass"
Format: speaker=<name>;<soundfile>;<X-Position>;<Y-Position>;<Z-Position>;<Volume>;[Start delay];[Keyframes];[Flags]
Example:
speaker=Speaker1;test.raw;1.0;1.0;1.75;1.0

Speaker keyframes: These allow movement of the speaker object and are a list of timestamps and
positions relative to the origin of the speaker. The engine interpolates the positions linearly.
A keyframe requires a time stamp and an x, y, z position, separated by a comma (not semicolon!).
Each keyframe is separated by a slash. Note that the rendering time increases when using movement
since the position has to be recalculated for every sound sample.
This example moves the speaker 3 meters to the right, 3 to the front and then 3 to the top:
speaker=Speaker1;test.raw;1.0;1.0;1.75;1.0;;1.0,3.0,0.0,0.0/3.5,3.0,3.0,0.0/6.3,3.0,3.0,3.0

The "microphone" receives sound waves and saves them as a mono audio file. It can be positioned
in threedimensional space. Up to 16 single microphones are currently allowed.
Format: microphone=<name><soundfile>;<X-Position>;<Y-Position>;<Z-Position>
Example:
microphone=Mic1;result.raw;0.0;0.0;1.7

The "2chanmic" saves a sound file with two channels and needs two threedimensional positions
as microphone positions.
Format: 2chanmic=<name>;<soundfile>;<Mic 1 X>;<Mic 1 Y>;<Mic 1 Z>;<Mic 2 X>;<Mic 2 Y>;<Mic 2 Z>
Example:
2chanmic=2ChanMic;output/output.raw;-0.08;0.0;1.75;0.08;0.0;1.75

A "head" is an object in threedimensional space that comes with two microphones as ears, can save a stereo
file and can be rotated from 0 to 360 degrees. An ear distance of 0.18 meters should deliver the most
realistic results.
Format: head=<name>;<soundfile>;<X-Position>;<Y-Position>;<Z-Position>;<Angle>;<Ear Distance>
Example:
head=Head;output/output.raw;1.0;0.0;1.7;180;0.18

A "wall" is a flat surface that reflects the sound to other walls and to microphones.
The entire soundwave is reflected when the amount value is at 1, non of the sound is reflected when the 
value is at 0. You can use any floating point value between 0 and 1. The wall needs two threedimensional
positions while one axis has to be the same values. Use the same x-values for side walls, the same y-
values for frontal walls and the same z-values for floors and ceilings. For realistic results, the
amount value should be pretty high, between 0.8 and 0.99.
Format: wall=<name>;<X-Position 1>;<Y-Position 1>;<Z-Position 1>;<X-Position 2>;<Y-Position 2>;<Z-Position 2>;<Amount>;[modifier]
Example (for a floor):
wall=Floor;-5.0;-1.0;0.0;5.0;6.0;0.0;0.9

The optional modifier value for a wall  is used if you don't want to have a perfectly straight wall,
which can lead to unrealistic smooth reverb in very big rooms or on big surfaces like house walls or
mountains. The surface can be modified by three different shapes: sine, square and triangle. You can
modify the length of these shapes and the depth (in meters). The pattern can be shifted by a given
amount of meters.
Format: <type>,<langth>,<depth>,<shift>
Example: wall=ModWall;-5.0;-1.0;0.0;5.0;6.0;0.0;0.9;square,10,3,0

A "room" creates 6 walls with identical reflection amounts. It requires two threedimensional positions which
should describe the lower left near and the upper right far point of a volume, as well as a reflection amount.
The reflection amount is a value between 0 and 1.
Format: room=<name>;<X-Position 1>;<Y-Position 1>;<Z-Position 1>;room=<X-Position 2>;<Y-Position 2>;<Z-Position 2>;<Amount>
Example:
room=Room;x-5.0;-1.0;0.0;5.0;6.0;3.0;0.9

The "rndspawnspeaker" is an advanced speaker with a lot of parameters that spawns with random delays in
random locations and plays in varying speeds. It can be used for water drops, birds, leaf sounds and so on.
It requires two positions that describe the corners of a cube in which the speakers spawn, a maximum
playback time (in seconds), the volume, the min/max pause time between spawns and the optional min/max
playback speed (1 is the default value).
Format: randomspawnspeaker=<name><file>;<X-Pos 1>;<Y-Pos 1>;<Z-Pos 1>;<X-Pos 2>;<Y-Pos 2>;<Z-Pos 2>;
        <Volume>;<Total Time>;<Min Pause>;<Max Pause>;[Min Speed];[Max Speed];[Flags]
Example:
randomspawnspeaker=RndSpeaker;test.raw;-10;0;0;10;20;5;1;60;3;6;0.8;1.2

"version" needs to be the first value in the file. it tells the engine, which file format should be used.
Current version is 0.5
Format: version=<version nr>

"nrbounces" is the number of reflection bounces between walls. The default value is 50. Smaller rooms
require more bounces than large rooms. Note that te render time will increase with the number of bounces.
Format: nrbounces=<number of bounces>

"nrreflections" determines the number of total reflection calculations before the engine finally saves the
microphone files and quits. The larger a room is, the more reflection passes are needed to sound well.
The total render time will increase with a rising number of reflections.
Format: nrreflections=<number of reflections>

"saveeverynrofpasses", when not set to zero, saves the microphone audiofiles after a given number of
reflection passes to offer the ability to hear the current result while still calculating and improving
the result. The rendering will continue after saving. Note that the total render time increases a bit
since the saving process takes some time.
Format: saveeverynrofpasses=<number of passes>

The "samplerate" value sets the sample rate of the input/output audio files. The default value is 44100.
Format: samplerate=<sample rate>

"additionaltime" sets an additional time to the output audio to allow echoes to complete. For a longer
reflection (in in big rooms or small rooms with many bounces and reflection amounts) a longer additional
audio time is required. The value is a floating point with a default value of 30 seconds.
Format: additionaltime=<additional time in seconds>

"randomseed" initiates the random number generator. The same seed will result in the same result
each time (same random spawn speakers, same reflectors). Changing the number will generate a new scene.
Default is 1.
Format: randomseed=<number>

"airdamping" is the amount of lowering high frequencies by higher distances between the sound source and
microphones. Reflections are dampened stronger since the sound waves travel a longer distance in total. A
high value results in lower frequency reduction, a low value in greater reduction. Default is 100.
Format: airdamping=<value>


changelog.txt
0.7
* GUI: Movement path and wall modifiers displayed in side/front perspective
* Fixed a bug where the engine took several GB of ram when a speaker path was invalid
* GUI: Movement Path Editor
* GUI: Optical improvements
* Minor improvements and fixes

0.6
* Engine: Added 'start delay' for speakers
* Engine: Added flags ('disabled', 'no direct', 'no indirect') for speakers and rndspeakers
* Engine: Fixed a bug that ignored reflection amount for moving speakers
* GUI: Added perspectives: front and side
* GUI: Added dialog for wall modifiers
* GUI: Changed colors
* GUI: Added fields for flags and start delay
* GUI: Added 'total play time' for rndspeakers
* GUI: Fixed windows bug that added new objects twice
* Small bug fixes and improvements


0.5
* Names for all objects
* version number
* Flags for speakers and random spawn speakers
* GUI editor (and console filter launcher)
* Wall groups
* Small changes and bug fixes

0.4
* High frequency damping of distant/reflective sounds
* New setting: airdamping
* Fixed a bug where direct samples were not saved
* Different bug fixes and code cleanup

0.3
* Wall modifiers (square, sine, triangle)
* Random spawn speakers
* Improved random number handling
* New setting: Random seed
* Minor improvements and bugfixes

0.2
* Speakers can now move by using keyframes

0.1
* First running version


src/GUI Editor/src/WallModEditor.cpp
#include "WallModEditor.h"

BEGIN_EVENT_TABLE(WallModEditorFrame, wxFrame)
    EVT_BUTTON(WallModEditorFrame::btnOkId, WallModEditorFrame::OnOkClick)
    EVT_TEXT(WallModEditorFrame::txtAmpId, WallModEditorFrame::OnTextChange)
    EVT_TEXT(WallModEditorFrame::txtWaveId, WallModEditorFrame::OnTextChange)
    EVT_TEXT(WallModEditorFrame::txtShiftId, WallModEditorFrame::OnTextChange)
    EVT_COMBOBOX(WallModEditorFrame::cmbTypeId, WallModEditorFrame::OnTextChange)

END_EVENT_TABLE()

WallModEditorFrame::WallModEditorFrame(wxFrame *frame, const wxString& title, BinauralProject *project, int objNr)
    : wxFrame(frame, -1, title, wxDefaultPosition, wxSize(250, 285), 0x800 | 0x400 | 0x200 | 0x1000 | wxCAPTION | wxCLIP_CHILDREN) {

    wxArrayString arrChoices;
    arrChoices.Add("None");
    arrChoices.Add("Sine");
    arrChoices.Add("Square");
    arrChoices.Add("Triangle");
    wxSize size = wxSize(this->GetSize().GetWidth() - 10, 20);
    new wxStaticText(this, wxID_ANY, "Modfier Type:", wxPoint(5, 0 * 50), size);
    comboType = new wxComboBox(this, cmbTypeId, "None", wxPoint(5, 0 * 50 + 20), size, arrChoices);
    new wxStaticText(this, wxID_ANY, "Wave length (m):", wxPoint(5, 1 * 50), size);
    editLength = new wxTextCtrl(this, txtWaveId, "", wxPoint(5, 1 * 50 + 20), size);
    new wxStaticText(this, wxID_ANY, "Amplitude (m):", wxPoint(5, 2 * 50), size);
    editAmp = new wxTextCtrl(this, txtAmpId, "", wxPoint(5, 2 * 50 + 20), size);
    new wxStaticText(this, wxID_ANY, "Shift (m):", wxPoint(5, 3 * 50), size);
    editShift = new wxTextCtrl(this, txtShiftId, "", wxPoint(5, 3 * 50 + 20), size);

    buttonOk = new wxButton(this, btnOkId, "OK", wxPoint(90, 4 * 50 + 20), wxSize(60, 20));

    myProject = project;
    currentObj = objNr;

    if (isObjectValid() == true) {
        origString = myProject->guiObjects[currentObj].modifier;
    }

    FillValues();
}

void WallModEditorFrame::FillValues() {
    if (isObjectValid() == true) {
        int usedMod = 0;
        double waveLength = 5.0, amplitude = 1.5, shift = 0.0;
        wxStringTokenizer token(myProject->guiObjects[currentObj].modifier, ",");

        if (token.HasMoreTokens()) {
            wxString nt = token.GetNextToken();
            if (nt == "sine") usedMod = 1;
            if (nt == "square") usedMod = 2;
            if (nt == "triangle") usedMod = 3;
        }
        if (token.HasMoreTokens())
            waveLength = atof(token.GetNextToken());
        if (token.HasMoreTokens())
            amplitude = atof(token.GetNextToken());
        if (token.HasMoreTokens())
            shift = atof(token.GetNextToken());

      blockEvents = true;
      editAmp->SetValue(fToStr(amplitude));
      editLength->SetValue(fToStr(waveLength));
      editShift->SetValue(fToStr(shift));

      comboType->SetSelection(usedMod);
      blockEvents = false;

    }
}

bool WallModEditorFrame::isObjectValid() {
   if (myProject != nullptr && currentObj >= 0 && currentObj < BinauralProject::MAX_OBJECTS
        && myProject->guiObjects[currentObj].objType == BinauralProject::OBJ_WALL) {

        return true;
    }

    return false;
}

WallModEditorFrame::~WallModEditorFrame() {
    if (clickedOk == false) {
        myProject->guiObjects[currentObj].modifier = origString;
        myProject->tempEvent.lastEvent = TempEvents::EVT_WM_CHANGED;
        myProject->tempEvent.objNr = currentObj;
    }
}

wxString WallModEditorFrame::generateString() {
    wxString ret;
    switch (comboType->GetSelection()) {
        case 0:
            return "";
        case 1:
            ret = "sine,";
            break;
        case 2:
            ret = "square,";
            break;
        case 3:
            ret = "triangle,";
            break;
    }
    ret += fToStr(atof(editLength->GetValue())) + ",";
    ret += fToStr(atof(editAmp->GetValue())) + ",";
    ret += fToStr(atof(editShift->GetValue()));

    return ret;
}

void WallModEditorFrame::OnOkClick(wxCommandEvent &evt) {
    clickedOk = true;
    Destroy();
    evt.Skip();
}

void WallModEditorFrame::OnTextChange(wxCommandEvent &evt) {
    if (isObjectValid() == true && blockEvents == false) {
        myProject->tempEvent.lastEvent = TempEvents::EVT_WM_CHANGED;
        myProject->tempEvent.objNr = currentObj;
        myProject->guiObjects[currentObj].modifier = generateString();
    }

    evt.Skip();
}
